src/libtls/tls_server.c

```diff
@@ -1350,6 +1359,57 @@ static status_t send_certificate_request(private_tls_server_t *this,
 	enumerator->destroy(enumerator);
 	writer->write_data16(writer, authorities->get_buf(authorities));
 	authorities->destroy(authorities);
+}
+
+/**
+ * Send Certificate Request
+ */
+static status_t send_certificate_request(private_tls_server_t *this,
+										 tls_handshake_type_t *type,
+										 bio_writer_t *writer)
+{
+	bio_writer_t *authorities, *supported, *extensions;
+
+	if (this->tls->get_version_max(this->tls) < TLS_1_3)
+	{
+		supported = bio_writer_create(4);
+		/* we propose both RSA and ECDSA */
+		supported->write_uint8(supported, TLS_RSA_SIGN);
+		supported->write_uint8(supported, TLS_ECDSA_SIGN);
+		writer->write_data8(writer, supported->get_buf(supported));
+		supported->destroy(supported);
+		if (this->tls->get_version_max(this->tls) >= TLS_1_2)
+		{
+			this->crypto->get_signature_algorithms(this->crypto, writer, TRUE);
+		}
+
+		write_certificate_authorities(writer);
+	}
+	else
+	{
+		/* certificate request context as described in RFC 8446, section 4.3.2 */
+		writer->write_uint8(writer, 0);
+
+		authorities = bio_writer_create(64);
+		write_certificate_authorities(authorities);
```

Comment:
Do you mean moving these two lines:

```
	authorities = bio_writer_create(64);
	write_certificate_authorities(authorities);
```

between the two extensions?

I think I moved this on purpose because of your comment a week ago:

> I'd make it the same as the others, i.e. remove CertificateRequest, because we see sending TLS CertificateRequest handshake in the log right afterwards (well, after the DNs for the CAs - adding them before the other extensions might clarify this even further).

Did I misunderstand you?

Reply:
I see, sorry about that. The code is just a bit strange this ways and the log messages are too (sending cert reqs, followed by the signature_algorithms extension and only then the certificate_authorities extension). I'd switch the order of the two extensions, so creation and use of the writer are closer together and the log would make a bit more sense (maybe if the cert_authorities extension is sent first, actually moving these lines after the "sending extension" log message would also be OK).

Reply:
I addressed this comment in the latest commits.